joomla/joomla-cms
# [4.0] [plg_system_httpheader] Implement security http headers per default and create the posibility to add custom http headers

### Summary of Changes

Implement security http headers per default and create the possibility to add custom http headers.

This plugin bases on https://github.com/zero-24/plg_system_httpheader which does the same for 3.x

### Testing Instructions

- test your webseite using: https://securityheaders.io/ => `The most if not all headers are missing`
- Install this PR.
- discover the new plugin
- check that the plugin is enabled if not enable it.
- check the report here: https://securityheaders.io/ again.

You can also check the http header using the `Network` Tab of your Browser:
![image](https://user-images.githubusercontent.com/2596554/31413039-d1d5e046-ae17-11e7-80b7-1b9db6cd11db.png)

Per default this plugin set the following headers + values:
```
X-Frame-Options: SAMEORIGIN
```
More Infos: https://scotthelme.co.uk/hardening-your-http-response-headers/#x-frame-options
```
X-XSS-Protection: 1; mode=block
```
More Infos: https://scotthelme.co.uk/hardening-your-http-response-headers/#x-xss-protection
```
X-Content-Type-Options: nosniff
```
More Infos: https://scotthelme.co.uk/hardening-your-http-response-headers/#x-content-type-options
```
Referrer-Policy: no-referrer-when-downgrade
```
More Infos: https://scotthelme.co.uk/a-new-security-header-referrer-policy/

### Expected result

After enabeling the default options the above mention headers are set.

### Actual result

Joomla does not set any security headers.

### Additional Infos

Give of the nature of the other headers we can not give a save default as they are very site specific. But atleast this plugin offers a easy to implement way to add the other headers using subforms.

Thanks for @SniperSister @wilsonge and @yvesh for the feedback on the first implementation ;)

### Next todo's

- [ ] Please test the plugin
- [x] We should review the language strings
- [x] We need to decide if the descriptions can stay or should be dropped. 
- [ ] Give your feedback.

Thanks.
